walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,212 @@
+      int numOfMandatoryEvents) {
+
+    // Generate all possible intervals in O(1) time
+    int totalMinutes = 60 * 24;
+    int meetingDuration = (int) request.getDuration();
+    ArrayList<TimeRange> possibleIntervals = new ArrayList<>();
+    for (int currentMinute = 0; currentMinute < totalMinutes - meetingDuration; currentMinute++) {
+      possibleIntervals.add(
+          TimeRange.fromStartEnd(currentMinute, currentMinute + meetingDuration, false));
+    }
+
+    // Maintain a frequency of timeSlot that already works for all mandatory attendee
+    // to the # of optional attendees
+    Map<TimeRange, Integer> frequency = new HashMap<>();
+
+    int startTime = 0;
+    int optimalEndTime = 0;
+    int prevOptionalAttendee = 0;
+    int prevMandatoryAttendee = 0;
+
+    // Check all possible intervals and return the optimal one with the most optional attendees
+    for (int j = 0; j < possibleIntervals.size(); j++) {
+      TimeRange curInterval = possibleIntervals.get(j);
+      int curMandatoryAttendee = 0;
+      int curOptionalAttendee = 0;
+      second:
+      for (Event i : events) {
+        // events do not overlap
+        if (!curInterval.overlaps(i.getWhen())) {
+          if (optionalAttendees.containsAll(i.getAttendees())) {
+            curOptionalAttendee++;
+          } else if (mandatoryAttendees.containsAll(i.getAttendees())) {
+            curMandatoryAttendee++;
+          }
+        } else {
+          // events overlap
+          if (mandatoryAttendees.containsAll(i.getAttendees())) {
+            break second;
+          }
+        }
+      }
+
+      // Initializes variables for first event
+      if (j == 0) {
+        startTime = curInterval.start();
+        optimalEndTime = curInterval.end();
+        prevOptionalAttendee = curOptionalAttendee;
+        prevMandatoryAttendee = curMandatoryAttendee;
+        continue;
+      }
+
+      // In order for overlapping intervals to be extended, previous and current intervals
+      // can fit all mandatory attendees and have the same number of optional attendee
+      if (curMandatoryAttendee == numOfMandatoryEvents
+          && curOptionalAttendee == prevOptionalAttendee) {
+
+        // Checks if this interval can be extended from the previous one
+        if (prevMandatoryAttendee == numOfMandatoryEvents) {
+          // Can merge with previous interval
+          optimalEndTime = curInterval.end();
+
+        } else {
+          // Not mergeable, must create a new interval
+          startTime = curInterval.start();
+          optimalEndTime = curInterval.end();
+        }
+      } else {
+        // Checks if the previous interval can be added
+        if (prevMandatoryAttendee == numOfMandatoryEvents) {
+          TimeRange slot = TimeRange.fromStartEnd(startTime, optimalEndTime, false);
+          frequency.put(slot, prevOptionalAttendee);
+        }
+        startTime = curInterval.start();
+        optimalEndTime = curInterval.end();
+      }
```

Comment:
This code would be easier to understand if you did the merge step *after* generating all the intervals, instead of while.

Reply:
I see, thanks. Can I leave the code as is? I somewhat forgot how I went about this algorithm.